File: networkmodule/src/main/java/com/test/networkmodule/NetworkSDK.java
package com.test.networkmodule;


import com.test.networkmodule.api.ShoppingAPI;
import com.test.networkmodule.api.ShoppingAPIImpl;
import com.test.networkmodule.di.DaggerNetworkComponent;
import com.test.networkmodule.di.NetworkComponent;
import com.test.networkmodule.exceptions.SDKNotInitializedException;

import javax.inject.Inject;

import dagger.Lazy;

public class NetworkSDK {

    private static final String TAG = "NetworkSDK";
    private static NetworkSDK INSTANCE;

    @Inject
    Lazy<ShoppingAPIImpl> shoppingAPILazy;

    private boolean isSDKInitialized;


    public static NetworkSDK getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new NetworkSDK();
        }
        return INSTANCE;
    }

    public void initialize() {
        NetworkComponent sdkComponent = DaggerNetworkComponent.builder()
                .build();
        sdkComponent.inject(this);
        isSDKInitialized = true;
    }

    public ShoppingAPI getShoppingAPI() {
        if (isSDKInitialized) {
            return shoppingAPILazy.get();
        } else {
            throw new SDKNotInitializedException();
        }
    }
}

File: networkmodule/src/main/java/com/test/networkmodule/NetworkSDK.kt
package com.test.networkmodule


import com.test.networkmodule.api.ShoppingAPI
import com.test.networkmodule.api.ShoppingAPIImpl
import com.test.networkmodule.di.DaggerNetworkComponent
import com.test.networkmodule.exceptions.SDKNotInitializedException
import dagger.Lazy
import javax.inject.Inject

class NetworkSDK {

    @Inject
    lateinit var shoppingAPILazy: Lazy<ShoppingAPIImpl>

    private var isSDKInitialized: Boolean = false

    val shoppingAPI: ShoppingAPI
        get() = when {
            isSDKInitialized -> shoppingAPILazy.get()
            else -> throw SDKNotInitializedException(
                "Network SDK is not initialized," +
                        "Please call the initialize method of SDK Manager"
            )
        }

    fun initialize() {
        val sdkComponent = DaggerNetworkComponent.builder()
            .build()
        sdkComponent.inject(this)
        isSDKInitialized = true
    }

    companion object {
        private var INSTANCE: NetworkSDK? = null

        val instance: NetworkSDK
            get() {
                when (INSTANCE) {
                    null -> INSTANCE = NetworkSDK()
                }
                return INSTANCE!!
            }
    }
}
